initial implementation of server

mozkeeler opened the pull request:
This is already deployed (and running) on the aws box, but here's the code (in server/). I'll keep improving it, but if you have any feedback, that'd be great.

mozkeeler committed:
initial commit of server implementation

monicachew commented on server/server.js:
```diff
@@ -0,0 +1,55 @@
+
+function createDB() {
+  console.log("creating if necessary...");
+  db = new sqlite3.Database("events.sqlite3", createTable);
```
is .sqlite3 a convention?

mozkeeler replied:
Oh - I guess just .sqlite is more common (at least, it's what firefox uses).

monicachew commented on server/server.js:
```diff
@@ -0,0 +1,55 @@
+        res.end("OK!");
+      });
+    });
+  }).listen(8158);
```
I am missing something, why is the port 8158?

mozkeeler replied:
Right - sorry. The webserver setup is a little weird. As far as I can tell, to be able to talk to blushproof.org:443, we actually run a server listening on :8443 of the box. So, we could just run a node server on port 8443. However, while I want to keep our key protected by a passphrase (I'll share it with you over a secure channel sometime), I don't want to have to enter it each time we start the node server. So, I set up an Nginx server to listen on 8443 as https and forward requests to port 8158 (8158 looks a little like "BLSH" (which, now that I look at it, looks a little like "bullshit", so maybe we should change that...)).

monicachew commented on server/server.js:
```diff
@@ -0,0 +1,55 @@
+               { ":timestamp" : evt.timestamp,
+                 ":event" : evt.event,
+                 ":userID" : userID,
+                 ":uploadID" : uploadID });
```
It would be good to document the json format somewhere in this file, i.e. at least that we extract the events blob for the JSON request (which also contains a bunch of other metadata)

I think we should also have a second timestamp corresponding to the time the blob was received. Otherwise we can't do things like sanity check that the upload interval is correct (e.g., the client-side timestamp should be at least PULSE_INTERVAL away from the server-side timestamp, and we shouldn't be receiving really old events)

monicachew replied:
And now I'm wondering if we need a version number, even though our format is really simple -- the scenario would be that later down the line we decide to change the format, and old clients still make requests to the upload server, so a version number is probably good to disambiguate.

monicachew replied:
I guess we can also defer the version numbering until the format changes -- at which point we can check for a non-existent version for alpha users if we need to.

mozkeeler replied:
I think a version number is a good idea. We could just pull the version number from the list of addons micropilot sends.

One thing I was thinking of that would obviate this a bit was to just dump the whole request into the database and do post-processing to get the data we actually cared about. However, then we're actually storing PII, so maybe this isn't a good idea.

monicachew replied:
I was thinking the version number of the data format, not the version number of FF. However, keeping the version number of FF is a good idea too.

Yeah, agreed on the not storing p2.

monicachew commented on server/server.js:
```diff
@@ -0,0 +1,55 @@
+      jsonstr += str;
+    });
+    req.on("end", function() {
+      var json = {};
```
I notice some people declare all their vars up front in the function as a reminder that of function scope -- ymmv, I dunno if that's a good idea or not.

monicachew commented on server/server.js:
```diff
@@ -0,0 +1,55 @@
+
+function createTable() {
+  console.log("creating table if necessary...");
+  db.run("CREATE TABLE IF NOT EXISTS events(timestamp INTEGER, event TEXT, userID TEXT, uploadID TEXT)", startServer);
```
This needs a comment that includes
- Units of the timestamp and the fact that it's rounded
- a pointer to the event types
- brief desc.of user and upload ids

mozkeeler committed:
improvements to server - no bp versioning, yet

mozkeeler commented:
Sorry - I guess re-basing a branch and then pushing to a remote messes things up. I think I addressed most of your comments, but I haven't implemented versioning the uploads yet.
